File: stress/metrics/gds.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Optional

from stress.measure.events import Event, EventType


@dataclass(frozen=True)
class GDSResult:
    gds: Optional[float]               # None = N/A
    n_levels: int
    stress_levels: List[float]
    completion_rates: List[float]
    monotonicity: Optional[float] = None  # Fraction of pairs with C_{i+1} <= C_i
    smoothness: Optional[float] = None    # Entropy-based step uniformity [0,1]
    na_reason: Optional[str] = None


def _compute_monotonicity(rates: List[float]) -> Optional[float]:
    """Fraction of adjacent pairs where C_{i+1} <= C_i. Range [0,1]."""
    if len(rates) < 2:
        return None
    pairs = len(rates) - 1
    mono = sum(1 for i in range(pairs) if rates[i + 1] <= rates[i])
    return mono / pairs


def _compute_smoothness(rates: List[float]) -> Optional[float]:
    """Entropy-based uniformity of degradation steps. Range [0,1].
    1.0 = perfectly uniform degradation. 0.0 = all degradation in one cliff-drop."""
    if len(rates) < 2:
        return None

    steps = [max(0.0, rates[i] - rates[i + 1]) for i in range(len(rates) - 1)]
    total_drop = sum(steps)

    if total_drop == 0.0:
        return 1.0  # No degradation = trivially smooth

    nonzero = [s for s in steps if s > 0.0]
    n_nonzero = len(nonzero)

    if n_nonzero <= 1:
        return 1.0 if len(steps) == 1 else 0.0

    p = [s / total_drop for s in nonzero]
    H = -sum(pi * math.log(pi) for pi in p)
    H_max = math.log(n_nonzero)

    return H / H_max if H_max > 0 else 1.0


def compute_gds(
    events: List[Event],
    expected_levels: Optional[List[float]] = None,
) -> GDSResult:
    """
    BP-1 — Graceful Degradation Score (GDS)

    Spec-aligned definition:
      Execute at ordered stress intensity levels s1..sn
      Measure completion rate Ci at each level
      GDS = (1/n) * Σ Ci

    Data source:
      We accept per-level completion evidence as events with:
        - type == WORK_UNIT_END (optional per-task) OR
        - events that directly carry (stress_level, completion_rate)
      For v0 reference implementation, we prefer the explicit evidence events:
        Event with completion_rate != None and stress_level != None.

    expected_levels:
      If provided, enforce that we have data for each declared level (orderable).
    """
    levels: List[float] = []
    rates: List[float] = []

    # Prefer explicit evidence events (stress_level + completion_rate)
    for e in events:
        if e.stress_level is not None and e.completion_rate is not None:
            levels.append(float(e.stress_level))
            rates.append(float(e.completion_rate))

    if not levels:
        return GDSResult(
            gds=None,
            n_levels=0,
            stress_levels=[],
            completion_rates=[],
            na_reason="No (stress_level, completion_rate) evidence found in events.",
        )

    # Basic bounds check
    for r in rates:
        if not (0.0 <= r <= 1.0):
            return GDSResult(
                gds=None,
                n_levels=len(levels),
                stress_levels=levels,
                completion_rates=rates,
                na_reason=f"completion_rate out of bounds: {r}",
            )

    # If expected levels are declared, enforce coverage
    if expected_levels is not None:
        exp = [float(x) for x in expected_levels]
        # Use tolerance-based comparison to avoid float equality issues
        missing = [
            x for x in exp
            if not any(math.isclose(x, g, rel_tol=1e-9) for g in levels)
        ]
        if missing:
            return GDSResult(
                gds=None,
                n_levels=len(levels),
                stress_levels=levels,
                completion_rates=rates,
                na_reason=f"Missing declared stress levels: {missing}",
            )

    n = len(rates)
    gds = sum(rates) / n
    gds = max(0.0, min(1.0, gds))

    # Ensure stress levels are monotonically orderable (spec requirement)
    # We don’t force monotonic in events ordering, but we ensure they can be sorted and disclosed.
    paired = sorted(zip(levels, rates), key=lambda x: x[0])
    s_sorted = [p[0] for p in paired]
    c_sorted = [p[1] for p in paired]

    mono = _compute_monotonicity(c_sorted)
    smooth = _compute_smoothness(c_sorted)

    return GDSResult(
        gds=gds,
        n_levels=n,
        stress_levels=s_sorted,
        completion_rates=c_sorted,
        monotonicity=mono,
        smoothness=smooth,
        na_reason=None,
    )

File: stress-ref/src/metrics/gds.rs
use crate::types::events::Event;

#[derive(Debug, Clone)]
pub struct GdsResult {
    pub gds: Option<f64>,
    pub n_levels: usize,
    pub stress_levels: Vec<f64>,
    pub completion_rates: Vec<f64>,
    pub monotonicity: Option<f64>,
    pub smoothness: Option<f64>,
    pub na_reason: Option<String>,
}

fn compute_monotonicity(rates: &[f64]) -> Option<f64> {
    if rates.len() < 2 { return None; }
    let pairs = rates.len() - 1;
    let mono = rates.windows(2).filter(|w| w[1] <= w[0]).count();
    Some(mono as f64 / pairs as f64)
}

fn compute_smoothness(rates: &[f64]) -> Option<f64> {
    if rates.len() < 2 { return None; }

    let steps: Vec<f64> = rates.windows(2)
        .map(|w| (w[0] - w[1]).max(0.0))
        .collect();
    let total_drop: f64 = steps.iter().sum();

    if total_drop == 0.0 { return Some(1.0); }

    let nonzero: Vec<f64> = steps.iter().copied().filter(|&s| s > 0.0).collect();
    let n_nonzero = nonzero.len();

    if n_nonzero <= 1 {
        return if steps.len() == 1 { Some(1.0) } else { Some(0.0) };
    }

    let p: Vec<f64> = nonzero.iter().map(|&s| s / total_drop).collect();
    let h: f64 = -p.iter().map(|&pi| pi * pi.ln()).sum::<f64>();
    let h_max = (n_nonzero as f64).ln();

    Some(if h_max > 0.0 { h / h_max } else { 1.0 })
}

/// BP-1: Graceful Degradation Score.
/// GDS = (1/n) * sum(C_i) where C_i is completion rate at stress level i.
pub fn compute_gds(events: &[Event], expected_levels: Option<&[f64]>) -> GdsResult {
    let mut levels = Vec::new();
    let mut rates = Vec::new();

    for e in events {
        if let (Some(sl), Some(cr)) = (e.stress_level, e.completion_rate) {
            levels.push(sl);
            rates.push(cr);
        }
    }

    if levels.is_empty() {
        return GdsResult {
            gds: None, n_levels: 0,
            stress_levels: vec![], completion_rates: vec![],
            monotonicity: None, smoothness: None,
            na_reason: Some("No (stress_level, completion_rate) evidence found".to_string()),
        };
    }

    for &r in &rates {
        if !(0.0..=1.0).contains(&r) {
            return GdsResult {
                gds: None, n_levels: levels.len(),
                stress_levels: levels, completion_rates: rates,
                monotonicity: None, smoothness: None,
                na_reason: Some(format!("completion_rate out of bounds: {r}")),
            };
        }
    }

    // Tolerance-based expected level enforcement
    if let Some(expected) = expected_levels {
        let missing: Vec<f64> = expected
            .iter()
            .filter(|&&exp| !levels.iter().any(|&got| (exp - got).abs() < 1e-9))
            .copied()
            .collect();
        if !missing.is_empty() {
            return GdsResult {
                gds: None, n_levels: levels.len(),
                stress_levels: levels, completion_rates: rates,
                monotonicity: None, smoothness: None,
                na_reason: Some(format!("Missing declared stress levels: {missing:?}")),
            };
        }
    }

    let n = rates.len() as f64;
    let gds = (rates.iter().sum::<f64>() / n).clamp(0.0, 1.0);

    // Sort by stress level for disclosure
    let mut paired: Vec<(f64, f64)> = levels.into_iter().zip(rates).collect();
    paired.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));

    let s_sorted: Vec<f64> = paired.iter().map(|p| p.0).collect();
    let c_sorted: Vec<f64> = paired.iter().map(|p| p.1).collect();
    let mono = compute_monotonicity(&c_sorted);
    let smooth = compute_smoothness(&c_sorted);

    GdsResult {
        gds: Some(gds),
        n_levels: s_sorted.len(),
        stress_levels: s_sorted,
        completion_rates: c_sorted,
        monotonicity: mono,
        smoothness: smooth,
        na_reason: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::events::{EventLog, EventType};

    #[test]
    fn perfect_gds() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for &level in &[0.1, 0.2, 0.3] {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(1.0);
        }
        log.emit(EventType::RunEnd);

        let result = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert_eq!(result.gds, Some(1.0));
    }

    #[test]
    fn degrading_gds() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[1.0, 0.5, 0.0]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);

        let result = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert!((result.gds.unwrap() - 0.5).abs() < 0.01);
    }

    #[test]
    fn no_evidence_na() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        log.emit(EventType::RunEnd);
        let result = compute_gds(log.events(), None);
        assert!(result.gds.is_none());
    }

    #[test]
    fn smoothness_uniform() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[1.0, 0.8, 0.6]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);
        let r = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert!((r.smoothness.unwrap() - 1.0).abs() < 0.01);
    }

    #[test]
    fn smoothness_cliff_drop() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[1.0, 1.0, 0.4]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);
        let r = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert!((r.smoothness.unwrap() - 0.0).abs() < 0.01);
    }

    #[test]
    fn monotonicity_decreasing() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[1.0, 0.8, 0.6]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);
        let r = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert_eq!(r.monotonicity, Some(1.0));
    }

    #[test]
    fn monotonicity_non_monotonic() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        for (&level, &rate) in [0.1, 0.2, 0.3].iter().zip(&[0.5, 0.8, 0.6]) {
            let e = log.emit(EventType::WorkUnitEnd);
            e.stress_level = Some(level);
            e.completion_rate = Some(rate);
        }
        log.emit(EventType::RunEnd);
        let r = compute_gds(log.events(), Some(&[0.1, 0.2, 0.3]));
        assert!((r.monotonicity.unwrap() - 0.5).abs() < 0.01);
    }
}
